lib/three_scale/middleware/presigned_downloads.rb

```diff
@@ -0,0 +1,88 @@
+# frozen_string_literal: true
+
+module ThreeScale
+  module Middleware
+    class PresignedDownloads
+      RE_SEPARATOR = ::File::SEPARATOR == "/" ? "/" : "[#{Regexp.escape(::File::SEPARATOR)}/]"
+
+      delegate *%i[requires_signing? good_signature? signature_at to_time], to: :class, private: true
+
+      class << self
+        def verifier
+          Rails.application.message_verifier("file downloads")
+        end
+
+        def requires_signing?(path)
+          # for the used separator see https://github.com/rack/rack/pull/2257
+          # question mark on first separator is paranoia of getting a relative path somehow
+          invoices_re = %r{#{RE_SEPARATOR}?system#{RE_SEPARATOR}.*#{RE_SEPARATOR}invoices#{RE_SEPARATOR}}
+          path.start_with? invoices_re
+        end
+
+        def sign_path_if_needed(path, expires_in = 3600)
+          requires_signing?(path) ? sign(path, expires_in) : path
+        end
+
+        def sign(path, expires_in = 3600)
+          sign_at(path, Time.now.utc.round + expires_in)
+        end
+
+        def sign_at(path, expires_at)
+          expires_timestamp = to_timestamp(expires_at)
+          signature = signature_at(path, expires_at)
+          params = { "3scale-Expires" => expires_timestamp, "3scale-Signature" => signature }.to_param
+
+          "#{path}?#{params}"
+        end
+
+        def signature_at(path, expires_at)
+          signing_options = { purpose: :download, expires_at: }
```

Comment:
wrt `file downloads` see https://api.rubyonrails.org/classes/ActiveSupport/MessageVerifier.html

wrt `expires_at`, this is not setting to `nil`, see https://bugs.ruby-lang.org/issues/14579

Also we don't ignore the `expired_at` parameter. Rather we use it to generate a signature initially as well on verification. And both signatures should match, assuring that on URL generation and on verification, the timestamp was the same.

btw `purpose` is not strictly needed, I just tohught it is good to add purpose, in case later we add other pre-signed URLs for other purposes. Although they may or may not use the same verifier.

Reply:
OK. So the initializer parameter is the `secret`. But if the project is open source, then the `secret` is public. Couldn't anyone generate a valid token knowing the `secret` and the `purpose`?

Reply:
I know you figured out. For any other readers though, this is just a salt when generating a new key. The underlying actual secret is `SECRET_KEY_BASE`